### PR TITLE
Extract ember related blueprints into ember proper

### PR DESCRIPTION
**Pull request by stefanpenner**
This aims to decouple features of ember-cli that are actually tied to ember, this should ease upgrades and also prevent the various blueprints from getting widely out of sync.

This lays the work that will eventually allow us transpose ember repo itself into a well formed add-on.
- [x] pull ember.js related blueprints into ember
- [x] decouple most of the blueprints from ember-cli
- [x] confirm all the blueprints already decoupled work
- [x] bundle shims (eventually they wont be needed at all)

Several pieces of blueprint functionality are still coupled to ember-cli, they depend on:
- [ ] decouple all blueprints from ember-cli
  - [x] ../../lib/valid-component-name
  - [ ] ../addon-import
  - [x] models/blueprint

There are some bootstraping issues, when it comes to `app` and `addon`themselves due to the cyclic nature of this flow. I do believe the simplest is for them to be part of this repo, but that assumption may prove to be wrong.
- [ ] deal with the cyclic nature of ember-cli -> ember -> app blueprint -> ember-cli, likely just by improving the mechanism by which we bootstrap the whole thing
- [x] `app.import` the various parts of `dist` (@rwjblue mind tackling this item?)
- [ ] blueprint test-harness so they can be tested outside of ember-cli themselves
- [ ] ... ?
- [ ] release ember as an npm module

### CODE REVIEW

**Comment by rwjblue** [comment]
The following blueprints should not be moved into Ember itself:

* `app` - This is the default folder structure/scaffolding for an app, and contains little or no Ember version specific information. 
* `addon` - Same as `app`.
* `adapter` - This is an Ember Data blueprint (not an Ember one).
* `model` - Ember Data
* `view` - This should likely be in `ember-legacy-views`

If we do need to defer some of the file contents for `app`/`addon` (perhaps things like `app/router.js`) we can make a custom blueprint for it (in `blueprints/router/`) and that can move into Ember...

**Comment by stefanpenner** [comment]
> app - This is the default folder structure/scaffolding for an app, and contains little or no Ember version specific information.
> addon - Same as app.

I am unclear about these. As parts of them are somewhat dictated
 by embers version and others are. For example.

**Comment by rwjblue** [comment]
> Maybe the right course of action for those, is to to split them apart.

Yep, agreed (I mentioned that in https://github.com/emberjs/ember.js/pull/12412#issuecomment-144505248).

**Comment by stefanpenner** [comment]
another thing is, the pod-layout vs not  pod-layout stuff is somewhat dictate by our ember docs. I suspect this is another case where we can parametrize.

File: blueprints/app/files/app/app.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import Ember from 'ember';
+import Resolver from 'ember-resolver';
+import loadInitializers from 'ember/load-initializers';
+import config from './config/environment';
+
+let App;
+
+Ember.MODEL_FACTORY_INJECTIONS = true;
```
Comment (stefanpenner) [inline blueprints/app/files/app/app.js]: @rwjblue for example this line (and this whole file) kinda wants to be in both

File: blueprints/app/files/bower.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "<%= name %>",
+  "dependencies": {
+    "ember": "2.0.0",
```
Comment (stefanpenner) [inline blueprints/app/files/bower.json]: As embers version changes, as do these dependencies. (once we drop bower though, that problem gets much nicer)

File: blueprints/app/files/config/environment.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/* jshint node: true */
+
+module.exports = function(environment) {
```
Comment (stefanpenner) [inline blueprints/app/files/config/environment.js]: this is totally something ember-cli specific

**Comment by stefanpenner** [comment]
Drawing the boundaries is kinda hard... hmm

File: blueprints/addon/files/ember-cli-build.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/* global require, module */
+var EmberAddon = require('ember-cli/lib/broccoli/ember-addon');
```
Comment (stefanpenner) [inline blueprints/addon/files/ember-cli-build.js]: i wonder if `ember-addon` and `ember-app` really want to be seperate modules... This hole NPM crap makes this so tricky

File: blueprints/app/files/app/app.js (hunk 1 above)
Comment (rwjblue) [inline blueprints/app/files/app/app.js]: Sure, as I mentioned in https://github.com/emberjs/ember.js/pull/12412#issuecomment-144505248 we will need to make a blueprint for this (and likely `router.js`) then invoke that blueprint from `app`.  The `app.js` blueprint lives in Ember, but the `app` structure blueprint lives in ember-cli...

File: index.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+var UnwatchedTree = require('broccoli-unwatched-tree');
+var mergeTrees = require('broccoli-merge-trees');
+
+module.exports = {
+  name: 'ember',
+  blueprintsPath: function() {
+    return path.join(__dirname, 'blueprints');
+  },
+
+  treeFor: function(type) {
```
Comment (rwjblue) [inline index.js]: Use `treeForVendor` instead of `treeFor` + guard for `type === 'vendor'`.

File: package.json [hunk 6]
```diff
@@ -24,6 +27,7 @@
     "bower": "~1.3.2",
     "chalk": "^0.5.1",
     "ember-cli": "^1.13.13",
+    "ember-cli-blueprint-test-helpers": "ember-cli/ember-cli-blueprint-test-helpers#qunit",
```
Comment (rwjblue) [inline package.json]: I'd like to land the changes from ember-cli/ember-cli-blueprint-test-helpers#qunit in a release version before merging this PR.

File: package.json [hunk 7]
```diff
@@ -31,6 +35,7 @@
     "emberjs-build": "0.5.3",
     "express": "^4.5.0",
     "finalhandler": "^0.4.0",
+    "fs-extra": "^0.26.3",
```
Comment (rwjblue) [inline package.json]: This is used from `blueprints/route/index.js` so it needs to be in `dependencies`.

**Comment by rwjblue** [comment]
This is looking very good, I left a few minor inline comments.  My main concerns at this point are:

* Should the `*-test` blueprints be moved into `ember-cli-qunit` instead of Ember itself?  It seems to me that this is the right choice and it would be unfortunate to have to move them twice (into Ember then out into ember-cli-qunit later).
* Can you provide a demo/test repo with the required setup to play with this?  How does it work with various ember-cli versions?
* We are defining our own shims here (which is good), does that have negative interaction on using ember-cli-shims also?  Which one wins if you have both?

**Comment by Turbo87** [comment]
> Should the *-test blueprints be moved into ember-cli-qunit instead of Ember itself? It seems to me that this is the right choice and it would be unfortunate to have to move them twice (into Ember then out into ember-cli-qunit later).

IMHO this project should use the same approach as taken in https://github.com/emberjs/data/pull/4167. i.e. providing both qunit and mocha blueprints from this repo.

**Comment by rwjblue** [comment]
Yes, agreed (we had come to the same conclusion but forgot to update here).

**Comment by Turbo87** [comment]
@rwjblue @stefanpenner I think this can be closed now that #13029 has been merged

**Comment by rwjblue** [comment]
Thanks for the reminder

**Comment by stefanpenner** [comment]
:+1:

**Pull request closed without merging**